irc/metadata.go

```diff
@@ -0,0 +1,257 @@
+package irc
+
+import (
+	"errors"
+	"regexp"
+	"slices"
+	"strings"
+
+	"maps"
+
+	"github.com/ergochat/ergo/irc/caps"
+	"github.com/ergochat/ergo/irc/modes"
+	"github.com/ergochat/ergo/irc/utils"
+)
+
+var (
+	errMetadataTooManySubs = errors.New("too many subscriptions")
+)
+
+type MetadataStore = map[string]string
+
+type MetadataThing = struct {
+	Target     string
+	Key        string
+	Visibility string // currently unused
+	Value      string
+}
+
+func metadataSet(server *Server, session *Session, target string, key string, value string) {
+	targetChannel := server.channels.Get(target)
+	targetClient := server.clients.Get(target)
+
+	if targetChannel != nil {
+		targetChannel.stateMutex.Lock()
+		defer targetChannel.stateMutex.Unlock()
+
+		targetChannel.metadata[key] = value
+		targetChannel.MarkDirty(IncludeAllAttrs)
+		return
+	}
+
+	if targetClient != nil {
+		targetClient.stateMutex.Lock()
+		defer targetClient.stateMutex.Unlock()
+
+		if targetClient.metadata == nil {
+			targetClient.metadata = make(MetadataStore)
+		}
+
+		targetClient.metadata[key] = value
+
+		// coming soon: https://www.youtube.com/watch?v=K14JkFfWUzc
+
+		return
+	}
+
+	notifySubscribers(server, session, target, key, value)
+}
+
+func notifySubscribers(server *Server, session *Session, target string, key string, value string) {
+	var notify utils.HashSet[*Session] = make(utils.HashSet[*Session])
+	targetChannel := server.channels.Get(target)
+	targetClient := server.clients.Get(target)
+
+	if targetClient != nil {
+		notify = targetClient.FriendsMonitors(caps.MetadataTwoJudgementDay)
+		// notify clients about changes regarding themselves
+		for _, s := range targetClient.Sessions() {
+			notify.Add(s)
+		}
+	}
+	if targetChannel != nil {
+		members := targetChannel.Members()
+		for _, m := range members {
+			for _, s := range m.Sessions() {
+				notify.Add(s)
```

Comment:
Move the cap check up here...

If you want to get adventurous you can actually use [iter.Seq](https://go.dev/blog/range-functions) here, so the client case will still allocate a hashset of friend sessions (because we need to deduplicate), but the channel case can just iterate over `Members()` and then `Sessions()` per member and return applicable sessions without allocating, and then `iter.Seq` is a common interface for ranging over both of these.